File: go/evm.go
package x402

import (
	"crypto/ecdsa"
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"math/big"
	"strings"
	"time"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/common/math"
	"github.com/ethereum/go-ethereum/crypto"
	"github.com/ethereum/go-ethereum/signer/core/apitypes"
)

// EVMSigner signs EIP-3009 ``TransferWithAuthorization`` messages with a
// locally-held ECDSA private key. Use NewEVMSignerFromPrivateKey to
// construct one.
type EVMSigner struct {
	privateKey *ecdsa.PrivateKey
	address    common.Address
}

// NewEVMSignerFromPrivateKey builds a signer from a hex-encoded secp256k1
// private key. The ``0x`` prefix is optional.
func NewEVMSignerFromPrivateKey(pkHex string) (*EVMSigner, error) {
	pkHex = strings.TrimPrefix(pkHex, "0x")
	pk, err := crypto.HexToECDSA(pkHex)
	if err != nil {
		return nil, fmt.Errorf("invalid EVM private key: %w", err)
	}
	return &EVMSigner{privateKey: pk, address: crypto.PubkeyToAddress(pk.PublicKey)}, nil
}

// Address returns the signer's checksummed 0x address.
func (s *EVMSigner) Address() string { return s.address.Hex() }

// randomNonce32 returns a 0x-prefixed random 32-byte nonce.
func randomNonce32() (string, error) {
	var b [32]byte
	if _, err := rand.Read(b[:]); err != nil {
		return "", err
	}
	return "0x" + hex.EncodeToString(b[:]), nil
}

// SignEVMPayment builds and signs a TransferWithAuthorization envelope
// for the given payment requirement. The returned envelope should be
// JSON-encoded and base64-encoded into the ``X-Payment`` header.
func SignEVMPayment(req PaymentRequirement, signer *EVMSigner) (*X402PaymentEnvelope, error) {
	if signer == nil {
		return nil, errors.New("x402: evm signer is nil")
	}
	if req.MaxAmountRequired == "" {
		return nil, errors.New("x402: maxAmountRequired is required")
	}
	value, ok := new(big.Int).SetString(req.MaxAmountRequired, 10)
	if !ok {
		return nil, fmt.Errorf("x402: invalid maxAmountRequired %q", req.MaxAmountRequired)
	}
	maxTimeout := req.MaxTimeoutSeconds
	if maxTimeout <= 0 {
		maxTimeout = 120
	}
	now := time.Now().Unix()
	nonce, err := randomNonce32()
	if err != nil {
		return nil, err
	}
	auth := EVMAuthorization{
		From:        signer.Address(),
		To:          req.PayTo,
		Value:       value.String(),
		ValidAfter:  fmt.Sprintf("%d", now),
		ValidBefore: fmt.Sprintf("%d", now+int64(maxTimeout)),
		Nonce:       nonce,
	}

	typedData, err := buildTypedData(req, auth)
	if err != nil {
		return nil, err
	}
	sig, err := signTypedData(signer.privateKey, typedData)
	if err != nil {
		return nil, fmt.Errorf("x402: EIP-712 sign: %w", err)
	}

	envelope := &X402PaymentEnvelope{
		X402Version: 2,
		Scheme:      firstNonEmpty(req.Scheme, "exact"),
		Network:     firstNonEmpty(req.Network, string(NetworkBase)),
		Payload:     EVMPayload{Authorization: auth, Signature: "0x" + hex.EncodeToString(sig)},
	}
	return envelope, nil
}

// buildTypedData mirrors typescript/src/evm.ts buildTypedData.
func buildTypedData(req PaymentRequirement, auth EVMAuthorization) (apitypes.TypedData, error) {
	extra := req.Extra
	if extra == nil {
		extra = map[string]any{}
	}

	domainName, _ := extra["name"].(string)
	if domainName == "" {
		domainName = "USD Coin"
	}
	domainVersion, _ := extra["version"].(string)
	if domainVersion == "" {
		domainVersion = "2"
	}

	// chainId may arrive as float64 (from JSON) or int.
	var chainID *big.Int
	switch v := extra["chainId"].(type) {
	case float64:
		chainID = big.NewInt(int64(v))
	case int:
		chainID = big.NewInt(int64(v))
	case int64:
		chainID = big.NewInt(v)
	case string:
		chainID, _ = new(big.Int).SetString(v, 10)
	}
	if chainID == nil {
		chainID = big.NewInt(8453) // Base mainnet default, matches TS.
	}

	verifyingContract, _ := extra["verifyingContract"].(string)
	if verifyingContract == "" {
		verifyingContract = req.Asset
	}
	if !common.IsHexAddress(verifyingContract) {
		return apitypes.TypedData{}, fmt.Errorf("x402: invalid verifyingContract %q", verifyingContract)
	}

	return apitypes.TypedData{
		Types: apitypes.Types{
			"EIP712Domain": []apitypes.Type{
				{Name: "name", Type: "string"},
				{Name: "version", Type: "string"},
				{Name: "chainId", Type: "uint256"},
				{Name: "verifyingContract", Type: "address"},
			},
			"TransferWithAuthorization": []apitypes.Type{
				{Name: "from", Type: "address"},
				{Name: "to", Type: "address"},
				{Name: "value", Type: "uint256"},
				{Name: "validAfter", Type: "uint256"},
				{Name: "validBefore", Type: "uint256"},
				{Name: "nonce", Type: "bytes32"},
			},
		},
		PrimaryType: "TransferWithAuthorization",
		Domain: apitypes.TypedDataDomain{
			Name:              domainName,
			Version:           domainVersion,
			ChainId:           math.NewHexOrDecimal256(chainID.Int64()),
			VerifyingContract: common.HexToAddress(verifyingContract).Hex(),
		},
		Message: apitypes.TypedDataMessage{
			"from":        auth.From,
			"to":          auth.To,
			"value":       auth.Value,
			"validAfter":  auth.ValidAfter,
			"validBefore": auth.ValidBefore,
			"nonce":       auth.Nonce,
		},
	}, nil
}

// signTypedData produces the 65-byte (R||S||V) EIP-712 signature. V is
// normalized to 27/28 to match Ethereum's canonical form, which is what
// the facilitator's ecrecover implementation expects.
func signTypedData(privKey *ecdsa.PrivateKey, typedData apitypes.TypedData) ([]byte, error) {
	domainSeparator, err := typedData.HashStruct("EIP712Domain", typedData.Domain.Map())
	if err != nil {
		return nil, fmt.Errorf("hash domain: %w", err)
	}
	messageHash, err := typedData.HashStruct(typedData.PrimaryType, typedData.Message)
	if err != nil {
		return nil, fmt.Errorf("hash message: %w", err)
	}
	raw := append([]byte{0x19, 0x01}, domainSeparator...)
	raw = append(raw, messageHash...)
	digest := crypto.Keccak256(raw)

	sig, err := crypto.Sign(digest, privKey)
	if err != nil {
		return nil, err
	}
	// crypto.Sign returns V as 0 or 1; canonical EIP-712 uses 27 or 28.
	sig[64] += 27
	return sig, nil
}

func firstNonEmpty(values ...string) string {
	for _, v := range values {
		if v != "" {
			return v
		}
	}
	return ""
}

File: typescript/src/evm.ts
/**
 * EVM X402 payment signing.
 *
 * Two schemes are supported and produce different envelopes:
 *
 * - `exact` — EIP-3009 `TransferWithAuthorization` against the token's own
 *   EIP-712 domain. Pre-authorises the exact amount in the 402 response.
 * - `upto`  — Permit2 `PermitWitnessTransferFrom` against the Permit2 EIP-712
 *   domain, with the AceData upto-proxy as spender. The signer authorises a
 *   ceiling; the facilitator settles for any amount ≤ that ceiling at /record
 *   time (zero settles with no on-chain transaction).
 *
 * Both work with any EIP-1193 provider (MetaMask, WalletConnect, etc.) — only
 * `eth_signTypedData_v4` is used.
 */

import type {
  EVMAuthorization,
  EVMPayload,
  EVMProvider,
  EVMUptoPayload,
  PaymentRequirement,
  Permit2WitnessAuthorization,
  X402PaymentEnvelope,
} from './types.js';

/** Canonical CREATE2 deployment of the Uniswap Permit2 contract. */
export const PERMIT2_ADDRESS = '0x000000000022D473030F116dDEE9F6B43aC78BA3';

/** Canonical CREATE2 deployment of the AceData upto-proxy spender. */
export const X402_UPTO_PERMIT2_PROXY_ADDRESS = '0x4020A4f3b7b90ccA423B9fabCc0CE57C6C240002';
const VALID_AFTER_SKEW_SECONDS = 30;

function randomNonce32(): string {
  const bytes = new Uint8Array(32);
  globalThis.crypto.getRandomValues(bytes);
  return '0x' + Array.from(bytes, (b) => b.toString(16).padStart(2, '0')).join('');
}

/** Uniformly random uint256 nonce for Permit2 (decimal string, no leading zeros). */
function randomPermit2Nonce(): string {
  const bytes = new Uint8Array(32);
  globalThis.crypto.getRandomValues(bytes);
  let value = 0n;
  for (const b of bytes) value = (value << 8n) | BigInt(b);
  return value.toString();
}

function buildTypedData(requirements: PaymentRequirement, authorization: EVMAuthorization) {
  return {
    types: {
      TransferWithAuthorization: [
        { name: 'from', type: 'address' },
        { name: 'to', type: 'address' },
        { name: 'value', type: 'uint256' },
        { name: 'validAfter', type: 'uint256' },
        { name: 'validBefore', type: 'uint256' },
        { name: 'nonce', type: 'bytes32' },
      ],
    },
    primaryType: 'TransferWithAuthorization' as const,
    domain: {
      name: requirements.extra?.name || 'USD Coin',
      version: requirements.extra?.version || '2',
      chainId: requirements.extra?.chainId || 8453,
      verifyingContract: requirements.extra?.verifyingContract || requirements.asset,
    },
    message: authorization,
  };
}

export async function signEVMPayment(
  requirements: PaymentRequirement,
  provider: EVMProvider,
  address: string
): Promise<X402PaymentEnvelope> {
  const now = Math.floor(Date.now() / 1000);
  // Default matches the official x402 SDK (3600s) and this package's upto path;
  // the server's 402 normally supplies the value.
  const maxTimeout = requirements.maxTimeoutSeconds || 3600;
  const value = BigInt(requirements.amount ?? requirements.maxAmountRequired ?? '0').toString();

  const authorization: EVMAuthorization = {
    from: address,
    to: requirements.payTo,
    value,
    validAfter: '0',
    validBefore: String(now + maxTimeout),
    nonce: randomNonce32(),
  };

  const typedData = buildTypedData(requirements, authorization);

  const signature = (await provider.request({
    method: 'eth_signTypedData_v4',
    params: [address, JSON.stringify(typedData)],
  })) as string;

  const payload: EVMPayload = { authorization, signature };

  return {
    x402Version: 2,
    accepted: requirements,
    payload,
  };
}

/**
 * Build the Permit2 `PermitWitnessTransferFrom` typed-data object.
 *
 * Mirrors the Python reference (`acedatacloud_x402.signing.evm._build_upto_typed_data`)
 * and the facilitator's `chain_handlers.upto_constants.build_upto_permit2_typed_data`
 * byte-for-byte: any drift breaks signature recovery.
 *
 * Critical:
 *   - The Permit2 EIP-712 domain has **no** `version` field. Adding one
 *     changes the digest and the facilitator recovers a different signer.
 *   - The witness struct is named `Witness` (NOT `X402Witness`).
 *   - The token's own domain is irrelevant — Permit2 holds the allowance.
 */
export function buildUptoTypedData(
  requirements: PaymentRequirement,
  params: {
    from: string;
    permittedAmount: bigint | string | number;
    nonce: bigint | string | number;
    deadline: bigint | string | number;
    validAfter: bigint | string | number;
  }
) {
  const extra = requirements.extra ?? {};
  const facilitatorAddress = extra.facilitatorAddress;
  if (!facilitatorAddress) {
    throw new Error(
      'PaymentRequirement.extra.facilitatorAddress is required for the upto scheme'
    );
  }
  const permit2Address = extra.permit2Address ?? PERMIT2_ADDRESS;
  const spender = extra.proxyAddress ?? X402_UPTO_PERMIT2_PROXY_ADDRESS;
  const verifyingContract = extra.verifyingContract ?? permit2Address;
  const chainId = extra.chainId ?? 8453;

  return {
    types: {
      PermitWitnessTransferFrom: [
        { name: 'permitted', type: 'TokenPermissions' },
        { name: 'spender', type: 'address' },
        { name: 'nonce', type: 'uint256' },
        { name: 'deadline', type: 'uint256' },
        { name: 'witness', type: 'Witness' },
      ],
      TokenPermissions: [
        { name: 'token', type: 'address' },
        { name: 'amount', type: 'uint256' },
      ],
      Witness: [
        { name: 'to', type: 'address' },
        { name: 'facilitator', type: 'address' },
        { name: 'validAfter', type: 'uint256' },
      ],
    },
    primaryType: 'PermitWitnessTransferFrom' as const,
    domain: {
      name: extra.name || 'Permit2',
      chainId,
      verifyingContract,
    },
    message: {
      permitted: { token: requirements.asset, amount: params.permittedAmount.toString() },
      spender,
      nonce: params.nonce.toString(),
      deadline: params.deadline.toString(),
      witness: {
        to: requirements.payTo,
        facilitator: facilitatorAddress,
        validAfter: params.validAfter.toString(),
      },
    },
  };
}

/**
 * Sign a Permit2 `PermitWitnessTransferFrom` envelope (`upto` scheme).
 *
 * `requirements.maxAmountRequired` is the **ceiling** the payer signs over.
 * The facilitator may settle any amount `≤` that ceiling at `/record` time.
 *
 * The signer must have already approved Permit2 to pull the token —
 * see {@link approvePermit2} for the one-time on-chain approval.
 */
export async function signEVMUptoPayment(
  requirements: PaymentRequirement,
  provider: EVMProvider,
  address: string,
  opts: { validAfter?: number; deadlineBuffer?: number; nonce?: bigint | string } = {}
): Promise<X402PaymentEnvelope> {
  const now = Math.floor(Date.now() / 1000);
  const validAfter = opts.validAfter ?? now - VALID_AFTER_SKEW_SECONDS;
  const timeout = opts.deadlineBuffer ?? requirements.maxTimeoutSeconds ?? 3600;
  const deadline = (opts.validAfter ?? now) + timeout;
  const nonce = opts.nonce ?? randomPermit2Nonce();
  const permittedAmount = BigInt(requirements.amount ?? requirements.maxAmountRequired ?? '0').toString();

  const typedData = buildUptoTypedData(requirements, {
    from: address,
    permittedAmount,
    nonce,
    deadline,
    validAfter,
  });

  const signature = (await provider.request({
    method: 'eth_signTypedData_v4',
    params: [address, JSON.stringify(typedData)],
  })) as string;

  const extra = requirements.extra ?? {};
  const spender = extra.proxyAddress ?? X402_UPTO_PERMIT2_PROXY_ADDRESS;
  const facilitatorAddress = extra.facilitatorAddress!;

  const permit2Authorization: Permit2WitnessAuthorization = {
    from: address,
    spender,
    nonce: nonce.toString(),
    deadline: deadline.toString(),
    permitted: { token: requirements.asset, amount: permittedAmount },
    witness: {
      to: requirements.payTo,
      facilitator: facilitatorAddress,
      validAfter: validAfter.toString(),
    },
  };

  const payload: EVMUptoPayload = { permit2Authorization, signature };

  return {
    x402Version: 2,
    accepted: requirements,
    payload,
  };
}

